### DISCUSSION TITLE
Footer alignment broke after Quarto 1.4.551

### DISCUSSION THREAD

**Discussion by dense-set** [quarto-dev/quarto-cli#9024] [Q&A] [answered]
### Description

I just updated from Quarto 1.4.3xx (don't remember the "xx") to 1.4.551. This has caused the alignment of the center and the right parts of the footer of my website (hosted on GitHub pages) to be left-aligned, which looks nasty. 

Before Quarto 1.4.551:
<img width="1443" alt="Screenshot 2024-03-11 at 5 55 26 AM" src="https://github.com/quarto-dev/quarto-cli/assets/102873841/a3b7174c-d2f7-48bc-a37b-04cc1b5f798a">

After Quarto 1.4.551:
<img width="1448" alt="Screenshot 2024-03-11 at 6 00 50 AM" src="https://github.com/quarto-dev/quarto-cli/assets/102873841/2e60d515-f737-4507-8d26-44a75842c3ab">

I haven't changed anything in the last commit that affects the footers. I also inspected the relevant footer sections on Firefox, and strangely that still shows the `text-align` to have correct values for `body > footer > div > div.nav-footer-x` for `x = left, center, right` (that is, `left`, `center` and `right` respectively). I am really confused as to what is happening here.

This is the [link](https://github.com/dense-set/dense-set.github.io) to my GitHub repo from which my [website](https://dense-set.github.io/) is being deployed (as guided on the Quarto documentation [here](https://quarto.org/docs/publishing/github-pages.html)).

Can someone have a look and please help me out?

---

The only sections in my repo relevant to the website footer are these:

- `_quarto.yml`: 
```yaml
format:
  html:
    css: common.css

website:
  page-footer: 
    border: true
    left: |
      [
        {{< iconify fa-brands creative-commons >}} 2024 Sarthak.
      ]{.footer-block}
      [
          All content licensed under<br>
          [{{< iconify fa-brands creative-commons >}}
          {{< iconify fa-brands creative-commons-by >}}
          {{< iconify fa-brands creative-commons-nc >}}
          {{< iconify fa-brands creative-commons-sa >}} CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en).
      ]{.footer-block}
    center: |
      *Be a hero,<br>
      or a villain even!;<br>
      only never just a mere prop.*
    right: |
      [
        Made with [Quarto](https://quarto.org/).
      ]{.footer-block}
      [
        View the source code on [GitHub {{< iconify fa-brands github >}}]({{< var website.repo-url >}}).
      ]{.footer-block}
```
- `common.css`:
```css
.footer-block {
  display: block;
  margin-bottom: 0.4 em;
}
```

**Comment by mcanouil** [comment]
Why are you doing this weird fenced div in footer?
Why not simply add a CSS rule (via `css` or more preferably via `theme`)?

Here, I assume your fenced/span div and CSS is broken rather than the footer from Quarto is broken.
Your rule needs to account for how  and what (S)CSS defines the footer.
I suggest you remove your fenced divs and (re)define (if really needed which I doubt) the CSS rather than modifying a global rule locally but everywhere.

Finally, don't mix HTML code with markdown code: `<br>` is two (or more) trailing spaces at the end in markdown.

**Reply by mcanouil** [reply]
More generally, it's better to modify the original rule rather than making local changes which in your case does not only add CSS class but also add an HTML div/span which means your content is more likely to break as you don't account for nested divs and CSS.

**Reply by dense-set** [reply]
I am trying to understand what you suggest. By "modify the original rule rather than making local changes", do you mean that it's preferable to use CSS selectors to directly target intended elements rather than first attaching classes to them and then targeting those classes?

**Reply by mcanouil** [reply]
Yes, as it is less likely to break over time.
Also, it avoid having several levels of CSS rules that interact with each other which in time will make it harder to debug.
In your case, there is no reason to use span/div, define a CSS class, provide that under `css` as opposed to provide the proper rule as SCSS under `theme` without requiring spans or divs, which means your YAML "footer" definition will also be way much "cleaner" and easier to read.

Note that this is mostly a personal recommandation from me. In the end, do what you think is best and works for you.

**Comment by cderv** [comment, marked as answer]
@dense-set 

Usually when this type of issue happens, especially if you are providing custom rules for CSS, you should enter browser debug mode, and try to understand which rules is creating the behavior you don't want. Possibly, either your own rule were overridden by a new rule from Quarto / Boostrap, or just something is not scoped enough or something else. 

In you case, I believe the issue is related to your CSS rule in `common.scss`
https://github.com/dense-set/dense-set.github.io/blob/2946b539542281487bfcba5bfd99ea295763f0b6/common.scss#L1-L3

````css
p {
    text-align: justify;
}
````

this will make all the `<p>` justify no matter where they are. If I opt out this rule in Browser Developer Tools I got back your expecting styling


![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/839ad0ac-3ba2-4bd2-83e3-65834f6c2e96)

I don't know exactly what could have change, but this rule in `common.scss` is really generic and applies to all `<p>` - you need to probably make it more specific (https://specifishity.com/)

I hope this helps understand how do look into those type of issue on your own. 

Also, some comment about what you do in your footer. 

````
      [
        {{< iconify fa-brands creative-commons >}} 2024 Sarthak.
      ]{.footer-block}
````

This syntax is for a Span. it will create a `<span>` in HTML with the class `footer-block`. This is ok to do, but I notice you are expecting this rule to apply 
````css
.footer-block {
  display: block;
  margin-bottom: 0.4 em;
}
````

`margin-bottom` is not a rule you can apply on a `<span>` so it will be ignored. The Developer Tools in browser shows you this too 

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/2b888f40-eefd-454b-a017-6c616a7714a9)

So either this is an oversight, you are not expecting a `<span>` but a `div` maybe. 

Anyhow, when trying to play with CSS, learning to use the Developer Tools in browser for Inspection is really valuable. There are a lot of resources in the web for this - starting point could be https://developer.chrome.com/docs/devtools/css

Hope this helps for your future work.

**Reply by dense-set** [reply]
That's it! Using 
```css
p:not(footer p) {
  text-align: justify;
}
```
fixed the issue!

Many thanks for the tips on better using DevTools. Yes, as I originally wrote, `margin-bottom` wasn't being applied, but not because it couldn't be applied; instead, I had a space between `0.4` and `em`. Fixing it did end up with the property being applied as expected.
